Wrappers/Python/cil/processors/PaganinProcessor.py

```diff
@@ -0,0 +1,537 @@
+            'magnification' : None,
+            'filter' : None,
+            'full_retrieval' : full_retrieval,
+            'pad' : pad,
+            'override_geometry' : None,
+            'override_filter' : None,
+            'return_multiplier' : return_multiplier
+            }
+        
+        super(PaganinProcessor, self).__init__(**kwargs)
+
+    def check_input(self, data):
+        if not isinstance(data, (AcquisitionData)):
+            raise TypeError('Processor only supports AcquisitionData')
+      
+        return True
+        
+    def process(self, out=None):
+
+        data  = self.get_input()
+
+        # set the geometry parameters to use from data.geometry unless the 
+        # geometry is overridden with an override_geometry
+        self._set_geometry(data.geometry, self.override_geometry)
+
+        if out is None:
+            out = data.geometry.allocate(None)
+
+        if self.pad>0:
+            try:
+                # pad the data
+                Padder.edge(pad_width={'horizontal': self.pad, 
+                                       'vertical':self.pad})(data)
+                Padder.edge(pad_width={'horizontal': self.pad, 
+                                       'vertical':self.pad})(out)
+            except:
+                # pad just in horizontal if the data is 2D
+                Padder.edge(pad_width={'horizontal': self.pad})(data)
+                Padder.edge(pad_width={'horizontal': self.pad})(out)
+
+        # create an empty axis if the data is 2D
+        if data.geometry.dimension == '2D':
+            data = np.expand_dims(data.as_array(),2)
+        else:
+            data = data.as_array()
+
+        # create a filter based on the shape of the data
+        filter_shape = np.shape(data.take(indices = 0, axis = 
+                                          out.get_dimension_axis('angle')))
+        self._create_filter(filter_shape[0], filter_shape[1], 
+                            self.override_filter)
+        
+        # pre-calculate the scaling factor
+        scaling_factor = -(1/self.mu)
+
+        # loop over the projections
+        for i in tqdm(range(len(out.geometry.angles))):
+            projection = data.take(indices = i, 
+                                   axis = out.get_dimension_axis('angle'))
+            if self.full_retrieval==True:
+                # apply the filter in fourier space, and scale by magnification
+                fI = fft2(self.magnification**2*projection)
+                iffI = ifft2(fI*self.filter)
+                # apply scaling factor
+                processed_projection = scaling_factor*np.log(iffI)
+            else:
+                # apply the filter in fourier space
+                fI = fft2(projection)
+                processed_projection = ifft2(fI*self.filter)
+            out.fill(np.squeeze(processed_projection), angle = i)
+        
+        # crop the data if it was previously padded
+        if self.pad>0:
```

Comment:
so if you use the padded buffers I suggested above you'd first copy a projection in to the buffer, run the filter and then copy back rather than slicing.